### ISSUE TITLE
FedAvgM

### ISSUE THREAD

**Issue by jafermarq** [adap/flower#2052]
# FedAvgM

* **Title:** Measuring the Effects of Non-Identical Data Distribution for Federated Visual Classification
* **Venue:** NeurIPSw 2021
* **Link to paper:** https://arxiv.org/abs/1909.06335

## Do you want to work on this baseline? 

> 🌻 Check everything about the `Summer of Reproducibility` on [`flower.dev/summer`](https://flower.dev/summer/)

> All available baselines are listed in the [Summer of Reproducibility Dashboard](https://github.com/orgs/adap/projects/6/) and also in the [GitHub Issues with the `summer-of-reproducibility` label](https://github.com/adap/flower/labels/summer-of-reproducibility). The content is the same.

> 📝 It is advised to complete these steps before your start working on your code. But if you can't wait to implement your baseline with Flower (we totally understand it 😄), please ensure you follow the steps on [how to contribute a new baseline](https://github.com/adap/flower/tree/main/baselines#contributing-a-new-baseline).

What follows are the steps 1 & 2 in the Summer of Reproducibility instructions.

### 1. Join the Summer of Reproducibility program
- [x] Join the [Flower Slack](https://flower.dev/join-slack/) and say _"hi! 👋"_ in the channel `#summer-of-reproducibility`.
- [x] Pick a baseline from our curated list **<---------------------------------------- [you are doing this now]**
### 2. Define the scope of your contribution
- [x] **What are you going to reproduce?** Add a comment to your issue and tell us about your plan regarding this baseline: what experiments from the paper are you reproducing?, for which datasets ? the more details you provide us with the better !
- [x] **Check if you are eligible for a reward**. 
    > As we have to comply with US/EU regulations, we have checked that individual contributors based on these countries or territories are eligible: Australia, Austria, Belgium, Bulgaria, Canada, Croatia, Cyprus, Czech Republic, Denmark, Estonia, Finland, France, Germany, Gibraltar, Greece, Hong Kong SAR China, Hungary, India, Ireland, Italy, Japan, Latvia, Liechtenstein, Lithuania, Luxembourg, Malta, Mexico, Netherlands, New Zealand, Norway, Poland, Portugal, Romania, Singapore, Slovakia, Slovenia, South Korea, Spain, Sweden, Switzerland, Thailand, United Arab emirates, United Kingdom United States.

    If where you are based is not on the list, please send us an email (`summer@flower.dev`) letting us know a bit about yourself (where are you currently based?, are you a university student? do you work at a public institution?). Please tell us the baselines you are interested in implementing (i.e. tell us your GitHub issue if you have crated one). We will reach back to you.
- [x] We will discuss with you about your contribution plan, if it sounds like a substantial enough contribution according to the Summer of Reproducibility rules (check our website [flower.dev/summer](https://flower.dev/summer/)), you'll get the OK from us to start working on your baseline!

## What happens next?

- [x] This item will be moved to the `In Progress` stage by a member of the Flower Team.

- [x] Follow the [instructions for creating a new baseline](https://github.com/adap/flower/tree/main/baselines#contributing-a-new-baseline) which will guide you through the process step-by-step.

## Is something wrong or not clear ?
* Ask a question directly in your issue.
* Reach out to us via the [Flower Slack](https://flower.dev/join-slack/) and ask your question in the #summer-of-reproducibility channel
* Check all the details (including FAQ) in the Summer of Reproducibility website: flower.dev/summer

**Comment by gubertoli** [comment]
Hi, I would like to implement this baseline. I had a look to the paper, as far as I understood the baseline needs to use the same data distribution proposed in the paper Figure 1 (considering CIFAR-10), and evaluates with the Momentum values {0, 0.7, 0.9, 0.97, 0.99, 0.997}. It seems that most of the work is to reproduce the data distribution, once the FedAvgM is already an available strategy on Flower.

**Comment by jafermarq** [comment]
Hey @gubertoli, good choice of baseline! Yes, as you mention the [`FedAvgM`](https://github.com/adap/flower/blob/main/src/py/flwr/server/strategy/fedavgm.py) strategy is already available in Flower. So you'd like to reproduce the results in Figure 5? and how about extending the same to FEMNIST (another popular dataset, even though is not included in this paper)? In this way there will be another opportunity for comparison with other baselines.

**Comment by gubertoli** [comment]
@jafermarq yes, adding FEMNIST seems ok.

**Comment by jafermarq** [comment]
@gubertoli  Great! this sounds like a plan then: CIFAR-10+FEMNIST. I have ✅ the first checks listed in the issue description. Could you please take a look if you'd be eligible for a reward before you start working on `FedAvgM`? If you have some questions please feel free to email us.

**Comment by gubertoli** [comment]
@jafermarq just confirming that I am based on one of the eligible countries (Germany).

**Comment by jafermarq** [comment]
@gubertoli , thanks for confirming. I've now ticked all the remaining points in Step 2 and I also made you the _assignee_ for this GitHub issue associated with `FedAvgM`. All set. If you click on the link shown in the _What happens next?_ section, it will present you a guide on how to start working on your baseline. Feel free to reach out to me if you encounter some problems or if you have questions (ideally in the Slack channel we have created for the _Summer of Reproducibility_)

Looking forward to seeing `FedAvgM` in action! 🙌
